File: Java/Gold/1005.java
import java.util.*;
import java.io.*;

public class Main {

    private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static ArrayList<ArrayList<Integer>> adj;
    private static ArrayList<ArrayList<Integer>> revAdj;
    private static HashMap<Integer, Boolean> isBuildTree;
    private static HashMap<Integer, Integer> depthMap;

    private static int W;
    private static int N;
    private static int[] costs;
    private static boolean[] visited;
    private static List<Integer> tpSortList;

    public static void main(String[] args) throws IOException{
        int T = Integer.parseInt(br.readLine());
        for (int i=0;i<T;i++) {
            settingInput();
            revDfs(W);
            dfsStartVertex();
            Collections.reverse(tpSortList);
            tpSortList.forEach( (vertex) -> {
                int beforeMaxCost = 0;
                ArrayList<Integer> beforeVertexes = revAdj.get(vertex);
                for (Integer beforeVertex : beforeVertexes) {
                    beforeMaxCost = beforeMaxCost >= depthMap.getOrDefault(beforeVertex, 0)
                            ? beforeMaxCost
                            : depthMap.getOrDefault(beforeVertex, 0);
                }
                depthMap.put(vertex, beforeMaxCost + costs[vertex]);
            });
            System.out.println(depthMap.get(W));
        }
        br.close();
    }

    private static void settingInput() throws IOException {
        adj = new ArrayList<>();
        revAdj = new ArrayList<>();
        isBuildTree = new HashMap<>();
        depthMap = new HashMap<>();
        tpSortList = new LinkedList<>();
        int[] NK = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        N = NK[0];
        int k = NK[1];
        visited = new boolean[N];
        for (int i = 0; i< N; i++) {
            adj.add(new ArrayList<>());
            revAdj.add(new ArrayList<>());
        }

        costs = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        for (int i = 0; i< k; i++) {
            String[] edge = br.readLine().split(" ");
            adj.get(Integer.parseInt(edge[0])-1).add(Integer.parseInt(edge[1])-1);
            revAdj.get(Integer.parseInt(edge[1])-1).add(Integer.parseInt(edge[0])-1);
        }
        W = Integer.parseInt(br.readLine())-1;
    }

    private static void revDfs(int vertex) {
        visited[vertex] = true;
        isBuildTree.putIfAbsent(vertex, true);

        revAdj.get(vertex).forEach( (next) -> {
            if (!visited[next]) {
                revDfs(next);
            }
        });

    }

    private static void dfsStartVertex() {
        visited = new boolean[N];
        for (int i=0;i<N;i++) {
            if (!visited[i] && isBuildTree.get(i) != null && revAdj.get(i).isEmpty()) {
                dfs(i);
            }
        }

    }

    private static void dfs(int vertex) {
        visited[vertex] = true;
        adj.get(vertex).forEach( next -> {
            if ( !visited[next] && isBuildTree.get(next) != null) {
                dfs(next);
            }
        });
        tpSortList.add(vertex);
    }
}

File: Archived2/Kotlin/Gold/1005.kt
import java.util.*;
import java.io.*;
import kotlin.collections.ArrayList

val br = BufferedReader(InputStreamReader(System.`in`))
var N = 0
var W = 0
lateinit var costs: List<Int>
lateinit var adj: ArrayList<ArrayList<Int>>
lateinit var revAdj: ArrayList<ArrayList<Int>>
lateinit var visited: BooleanArray
lateinit var isBuildTree: HashMap<Int, Boolean>
lateinit var tpSortList: MutableList<Int>

fun main() {
    val time = br.readLine().toInt()
    repeat(time) {
        settingInput()
        settingBuildTree()
        dfsAll()
        tpSortList.reverse()
        val costMap = HashMap<Int, Int>()
        tpSortList.forEach { vertex ->
            var vertexMaxCost = 0
            revAdj[vertex].forEach {before ->
                vertexMaxCost =
                    if (vertexMaxCost >= costMap.getOrDefault(before, 0)) vertexMaxCost
                    else costMap.getOrDefault(before,0)
            }
            costMap.put(vertex, vertexMaxCost+costs[vertex])
        }
        println(costMap[W])
    }
    br.close()
}

fun settingInput() {
    val NM = br.readLine().split(" ").map { it.toInt() }
    N = NM[0]
    val M = NM[1]
    costs = br.readLine().split(" ").map { it.toInt() }
    visited = BooleanArray(N)
    adj = ArrayList()
    revAdj = ArrayList()
    repeat(N) {
        adj.add(ArrayList())
        revAdj.add(ArrayList())
    }
    repeat(M) {
        val edge = br.readLine().split(" ").map { it.toInt() }
        adj.get(edge[0]-1).add(edge[1]-1)
        revAdj.get(edge[1]-1).add(edge[0]-1)
    }
    W = br.readLine().toInt() -1
}

fun settingBuildTree() {
    isBuildTree = HashMap()
    reverseDfs(W)
}

fun reverseDfs(vertex: Int) {
    visited[vertex] = true
    isBuildTree.putIfAbsent(vertex, true)
    revAdj[vertex].forEach { next ->
        if (!visited[next]) {
            reverseDfs(next)
        }
    }
}

fun dfsAll() {
    tpSortList = mutableListOf()
    visited = BooleanArray(N)
    repeat(N) { vertex ->
        if (isBuildTree[vertex] != null && revAdj[vertex].isEmpty() && !visited[vertex]) {
            dfs(vertex)
        }
    }
}

fun dfs(vertex: Int) {
    visited[vertex] = true
    adj[vertex].forEach { next ->
        if (!visited[next]) {
            dfs(next)
        }
    }
    tpSortList.add(vertex)
}